File: lib/component/newlib/binding.c
#include <errno.h>
#include <fcntl.h>
#include <stdint.h>
#include <sys/poll.h>
#include <sys/stat.h>
#include <time.h>
#include <unistd.h>

int espz_newlib_errno(void)
{
    return errno;
}

int espz_newlib_poll(struct pollfd *fds, uint32_t nfds, int32_t timeout)
{
    return poll(fds, (nfds_t)nfds, timeout);
}

void espz_newlib_close(int fd)
{
    (void)close(fd);
}

int espz_newlib_open(const char *path, int flags, uint32_t mode)
{
    return open(path, flags, (mode_t)mode);
}

int espz_newlib_fcntl(int fd, int cmd, uintptr_t arg)
{
    switch (cmd) {
        case F_GETFL:
            return fcntl(fd, cmd);
        case F_SETFL:
            return fcntl(fd, cmd, (int)arg);
        default:
            errno = EINVAL;
            return -1;
    }
}

ssize_t espz_newlib_read(int fd, void *buf, size_t len)
{
    return read(fd, buf, len);
}

ssize_t espz_newlib_write(int fd, const void *buf, size_t len)
{
    return write(fd, buf, len);
}

int64_t espz_newlib_lseek_set(int fd, uint64_t offset)
{
    return (int64_t)lseek(fd, (off_t)offset, SEEK_SET);
}

int64_t espz_newlib_lseek_cur(int fd, int64_t offset)
{
    return (int64_t)lseek(fd, (off_t)offset, SEEK_CUR);
}

int64_t espz_newlib_lseek_cur_get(int fd)
{
    return (int64_t)lseek(fd, (off_t)0, SEEK_CUR);
}

int64_t espz_newlib_lseek_end(int fd, int64_t offset)
{
    return (int64_t)lseek(fd, (off_t)offset, SEEK_END);
}

int espz_newlib_mkdir(const char *path, uint32_t mode)
{
    return mkdir(path, (mode_t)mode);
}

int espz_newlib_unlink(const char *path)
{
    return unlink(path);
}

int espz_newlib_clock_gettime_monotonic(struct timespec *ts)
{
    return clock_gettime(CLOCK_MONOTONIC, ts);
}

File: lib/component/newlib/binding.zig
pub const fd_t = i32;
pub const mode_t = u32;
const time_t = i64;

pub const timeval = extern struct {
    tv_sec: time_t,
    tv_usec: c_long,
};

pub const timespec = extern struct {
    tv_sec: time_t,
    tv_nsec: c_long,
};

pub const pollfd = extern struct {
    fd: fd_t,
    events: c_short,
    revents: c_short,
};

pub extern fn espz_newlib_errno() c_int;
pub extern fn espz_newlib_poll(fds: ?*pollfd, nfds: u32, timeout: i32) c_int;
pub extern fn espz_newlib_close(fd: fd_t) void;
pub extern fn espz_newlib_open(path: [*:0]const u8, flags: c_int, mode: mode_t) c_int;
pub extern fn espz_newlib_fcntl(fd: fd_t, cmd: c_int, arg: usize) c_int;
pub extern fn espz_newlib_read(fd: fd_t, buf: ?*anyopaque, len: usize) isize;
pub extern fn espz_newlib_write(fd: fd_t, buf: ?*const anyopaque, len: usize) isize;
pub extern fn espz_newlib_lseek_set(fd: fd_t, offset: u64) i64;
pub extern fn espz_newlib_lseek_cur(fd: fd_t, offset: i64) i64;
pub extern fn espz_newlib_lseek_cur_get(fd: fd_t) i64;
pub extern fn espz_newlib_lseek_end(fd: fd_t, offset: i64) i64;
pub extern fn espz_newlib_mkdir(path: [*:0]const u8, mode: mode_t) c_int;
pub extern fn espz_newlib_unlink(path: [*:0]const u8) c_int;
pub extern fn espz_newlib_clock_gettime_monotonic(ts: *timespec) c_int;
